File: apps/web-ui/src/lib/api/studio-sessions.ts
import { serviceEndpoints } from '@/lib/service-endpoints'

export type StudioSessionStatus =
  | 'provisioning'
  | 'ready'
  | 'busy'
  | 'idle'
  | 'degraded'
  | 'stopping'
  | 'stopped'
  | 'failed'
  | 'expired'

export type StudioRuntimeProfile = 'standard' | 'high_mem' | 'gpu'

export type WorkspaceLaunchMode = 'reuse_active_runtime' | 'provision_new_runtime'

export interface StudioSession {
  id: string
  project_id: string
  owner_user_id: string
  display_name: string
  runtime_profile_id: StudioRuntimeProfile
  runtime_session_id: string
  assistant_session_id: string
  status: StudioSessionStatus
  metadata: Record<string, unknown>
  created_at: string
  updated_at: string
  last_activity_at: string
}

export interface WorkspaceHandoff {
  project_id: string
  runtime_session_id: string | null
  runtime_profile_id: StudioRuntimeProfile
  launch_mode: WorkspaceLaunchMode
  workspace_url: string
  target_path: string | null
  notebook_path: string | null
  open_artifact_id: string | null
  initial_focus: string | null
  materialize_notebook_if_needed: boolean
}

export interface CreateStudioSessionRequest {
  project_id: string
  display_name: string
  runtime_profile_id?: StudioRuntimeProfile
  attach_if_exists?: boolean
  metadata?: Record<string, unknown>
}

export interface ListStudioSessionsOptions {
  project_id?: string
  runtime_profile_id?: StudioRuntimeProfile
  status?: StudioSessionStatus
  limit?: number
  offset?: number
}

export interface StudioSessionActionRequest {
  reason?: string
}

export interface WorkspaceHandoffRequest {
  runtime_profile_id?: StudioRuntimeProfile
  target_path?: string | null
  notebook_path?: string | null
  open_artifact_id?: string | null
  initial_focus?: string | null
  materialize_notebook_if_needed?: boolean
  open_clean_workspace?: boolean
}

type RequestOptions = {
  accessToken?: string | null
  signal?: AbortSignal
}

const STUDIO_SESSIONS_BASE = '/api/studio/sessions'

const buildHeaders = (accessToken?: string | null) => {
  const headers: HeadersInit = {
    'Content-Type': 'application/json',
  }
  if (accessToken) {
    headers.Authorization = `Bearer ${accessToken}`
  }
  return headers
}

async function requestJson<T>(
  path: string,
  init: RequestInit,
  options?: RequestOptions,
): Promise<T> {
  const response = await fetch(serviceEndpoints.orchestrator(path), {
    ...init,
    credentials: 'include',
    signal: options?.signal,
    headers: {
      ...buildHeaders(options?.accessToken),
      ...(init.headers ?? {}),
    },
  })

  if (!response.ok) {
    let detail = `${response.status} ${response.statusText}`
    try {
      const payload = (await response.json()) as { detail?: string }
      if (payload?.detail) {
        detail = payload.detail
      }
    } catch {
      // Ignore non-JSON failures and surface the HTTP status text instead.
    }
    throw new Error(`Studio session gateway request failed: ${detail}`)
  }

  return (await response.json()) as T
}

export async function createOrAttachStudioSession(
  payload: CreateStudioSessionRequest,
  options?: RequestOptions,
): Promise<StudioSession> {
  const response = await requestJson<{ session: StudioSession }>(
    STUDIO_SESSIONS_BASE,
    {
      method: 'POST',
      body: JSON.stringify(payload),
    },
    options,
  )
  return response.session
}

export async function listStudioSessions(
  params: ListStudioSessionsOptions = {},
  options?: RequestOptions,
): Promise<StudioSession[]> {
  const query = new URLSearchParams()
  if (params.project_id) query.set('project_id', params.project_id)
  if (params.runtime_profile_id) query.set('runtime_profile_id', params.runtime_profile_id)
  if (params.status) query.set('status', params.status)
  if (typeof params.limit === 'number') query.set('limit', String(params.limit))
  if (typeof params.offset === 'number') query.set('offset', String(params.offset))
  const path = query.size > 0 ? `${STUDIO_SESSIONS_BASE}?${query.toString()}` : STUDIO_SESSIONS_BASE

  const response = await requestJson<{ items: StudioSession[] }>(
    path,
    { method: 'GET' },
    options,
  )
  return response.items
}

export async function getStudioSession(
  sessionId: string,
  options?: RequestOptions,
): Promise<StudioSession> {
  const response = await requestJson<{ session: StudioSession }>(
    `${STUDIO_SESSIONS_BASE}/${encodeURIComponent(sessionId)}`,
    { method: 'GET' },
    options,
  )
  return response.session
}

export async function performStudioSessionAction(
  sessionId: string,
  action: 'touch' | 'close',
  payload: StudioSessionActionRequest = {},
  options?: RequestOptions,
): Promise<StudioSession> {
  const response = await requestJson<{ action: string; session: StudioSession }>(
    `${STUDIO_SESSIONS_BASE}/${encodeURIComponent(sessionId)}/actions/${action}`,
    {
      method: 'POST',
      body: JSON.stringify(payload),
    },
    options,
  )
  return response.session
}

export async function buildWorkspaceHandoff(
  sessionId: string,
  payload: WorkspaceHandoffRequest = {},
  options?: RequestOptions,
): Promise<WorkspaceHandoff> {
  const response = await requestJson<{ handoff: WorkspaceHandoff }>(
    `${STUDIO_SESSIONS_BASE}/${encodeURIComponent(sessionId)}/workspace-handoff`,
    {
      method: 'POST',
      body: JSON.stringify(payload),
    },
    options,
  )
  return response.handoff
}

File: packages/brain-researcher/src/brain_researcher/services/orchestrator/endpoints/studio_sessions.py
"""FastAPI endpoints for the hosted Studio session gateway.

Every session-scoped endpoint in this router must return 404, not 403,
when the caller is not the session owner. That prevents leaking whether a
given opaque session id exists for another user.
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, HTTPException, Query, Request

from ..studio_session_runtime import (
    CreateStudioSessionRequest,
    StudioRuntimeProfile,
    StudioSessionActionRequest,
    StudioSessionRuntime,
    StudioSessionStatus,
    WorkspaceHandoffRequest,
)

router = APIRouter(prefix="/api/studio/sessions", tags=["studio-sessions"])


def _runtime_from_request(request: Request) -> StudioSessionRuntime:
    runtime = getattr(request.app.state, "studio_session_runtime", None)
    if runtime is None:
        raise HTTPException(
            status_code=503, detail="Studio session runtime is not available"
        )
    return runtime


async def _resolve_request_user(request: Request):
    from ..auth_endpoints import _resolve_authenticated_user

    return await _resolve_authenticated_user(request)


@router.post("")
async def create_or_attach_studio_session(
    request: Request, payload: CreateStudioSessionRequest
) -> dict[str, Any]:
    runtime = _runtime_from_request(request)
    user, _ = await _resolve_request_user(request)
    session = await runtime.create_or_attach_session(user.id, payload)
    return {"session": session.model_dump(mode="json")}


@router.get("")
async def list_studio_sessions(
    request: Request,
    project_id: str | None = None,
    runtime_profile_id: StudioRuntimeProfile | None = None,
    status: StudioSessionStatus | None = None,
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
) -> dict[str, Any]:
    runtime = _runtime_from_request(request)
    user, _ = await _resolve_request_user(request)
    items = await runtime.list_sessions(
        owner_user_id=user.id,
        project_id=project_id,
        runtime_profile_id=runtime_profile_id,
        status=status,
        limit=limit,
        offset=offset,
    )
    return {"items": [item.model_dump(mode="json") for item in items]}


@router.get("/{session_id}")
async def get_studio_session(request: Request, session_id: str) -> dict[str, Any]:
    runtime = _runtime_from_request(request)
    user, _ = await _resolve_request_user(request)
    session = await runtime.get_session(session_id)
    if session is None or session.owner_user_id != user.id:
        raise HTTPException(status_code=404, detail="Studio session not found")
    return {"session": session.model_dump(mode="json")}


@router.post("/{session_id}/actions/{action}")
async def perform_studio_session_action(
    request: Request,
    session_id: str,
    action: str,
    payload: StudioSessionActionRequest | None = None,
) -> dict[str, Any]:
    runtime = _runtime_from_request(request)
    user, _ = await _resolve_request_user(request)
    try:
        return await runtime.perform_action(user.id, session_id, action, payload)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail="Studio session not found") from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=f"Unsupported action: {action}") from exc


@router.post("/{session_id}/workspace-handoff")
async def create_workspace_handoff(
    request: Request,
    session_id: str,
    payload: WorkspaceHandoffRequest,
) -> dict[str, Any]:
    runtime = _runtime_from_request(request)
    user, _ = await _resolve_request_user(request)
    try:
        handoff = await runtime.build_workspace_handoff(user.id, session_id, payload)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail="Studio session not found") from exc
    return {"handoff": handoff.model_dump(mode="json")}
